src/services/services.ts

```diff
@@ -6259,6 +6373,187 @@ namespace ts {
+                        getSymbolsForComponentTypes(<UnionOrIntersectionType>componentType, result);
+                    }
+                }
+                return result;
+            }
+
+            function getContainingTypeReference(node: Node): Node {
+                let topLevelTypeReference: Node = undefined;
+
+                while (node) {
+                    if (isTypeNode(node)) {
+                        topLevelTypeReference = node;
+                    }
+                    node = node.parent;
+                }
+
+                return topLevelTypeReference;
+            }
+
+            function getContainingClassHeritageClause(node: Node): HeritageClause {
+                if (node) {
+                    if (node.kind === SyntaxKind.ExpressionWithTypeArguments
+                        && node.parent.kind === SyntaxKind.HeritageClause
+                        && isClassLike(node.parent.parent)) {
+                        return <HeritageClause>node.parent;
+                    }
+
+                    else if (node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.PropertyAccessExpression) {
+                        return getContainingClassHeritageClause(node.parent);
+                    }
+                }
+                return undefined;
+            }
+
+            /**
+             * Returns true if this is an expression that could be used to implement an interface.
+             */
+            function isImplementationExpression(node: Expression): boolean {
+                // Unwrap parentheses
+                if (node.kind === SyntaxKind.ParenthesizedExpression) {
+                    return isImplementationExpression((<ParenthesizedExpression>node).expression);
+                }
+
+                return node.kind === SyntaxKind.ArrowFunction ||
+                    node.kind === SyntaxKind.FunctionExpression ||
+                    node.kind === SyntaxKind.ObjectLiteralExpression ||
+                    node.kind === SyntaxKind.ClassExpression ||
+                    node.kind === SyntaxKind.ArrayLiteralExpression;
+            }
+
+            /**
+             * Determines if the parent symbol occurs somewhere in the child's ancestry. If the parent symbol
+             * is an interface, determines if some ancestor of the child symbol extends or inherits from it.
+             * This also takes in a cache of previous results which makes this slightly more efficient and is
+             * necessary to avoid potential loops like so:
+             *     class A extends B { }
+             *     class B extends A { }
+             *
+             * @param child         A class or interface Symbol
+             * @param parent        Another class or interface Symbol
+             * @param cachedResults A map of symbol names to booleans indicating previous results
+             */
+            function inheritsFrom(child: Symbol, parent: Symbol, cachedResults: Map<boolean>): boolean {

```

Comment:
Thanks for the explanation! Could you move that into the doc comment? Could also rename the function to `explicitlyInheritsFrom`.